dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/emitter/nack/KafkaEmitterNackRestfulTest.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+
+import static com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions.SERVER_ONLY;
+import static com.ibm.ws.microprofile.reactive.messaging.fat.kafka.common.KafkaUtils.kafkaClientLibs;
+import static com.ibm.ws.microprofile.reactive.messaging.fat.kafka.common.KafkaUtils.kafkaPermissions;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertNotNull;
+
+@RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
+public class KafkaEmitterNackRestfulTest {
+
+    public static final String APP_NAME = "kafka-message-restful-emitter";
+    public static final String SERVER_NAME = "SimpleRxMessagingServer";
+
+    @Server(SERVER_NAME)
+    public static LibertyServer server;
+
+    @ClassRule
+    public static final RepeatTests r = ReactiveMessagingActions.repeat(SERVER_NAME, ReactiveMessagingActions.MP61_RM30, ReactiveMessagingActions.MP50_RM30,ReactiveMessagingActions.MP60_RM30);
+
+    @BeforeClass
+    public static void setup() throws Exception {
+
+        Map<String, Object> invalidKafkaConfig = Collections.singletonMap(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "PLAINTEXT://localhost:10000");
+
+        PropertiesAsset appConfig = new PropertiesAsset()
+                .addProperty(AbstractKafkaTestServlet.KAFKA_BOOTSTRAP_PROPERTY, KafkaTests.kafkaContainer.getBootstrapServers())
+                .include(ConnectorProperties.simpleOutgoingChannel(invalidKafkaConfig, EmitterRestResource.PAYLOAD_CHANNEL_NAME))
+                .include(ConnectorProperties.simpleOutgoingChannel(invalidKafkaConfig, EmitterRestResource.MESSAGE_CHANNEL_NAME));
+
+
+        WebArchive war = ShrinkWrap.create(WebArchive.class, APP_NAME + ".war")
+                .addAsLibraries(kafkaClientLibs())
+                .addAsManifestResource(kafkaPermissions(), "permissions.xml")
+                .addPackage(EmitterApplication.class.getPackage())
+                .addPackage(KafkaTestConstants.class.getPackage())
+                .addAsResource(appConfig, "META-INF/microprofile-config.properties");
+
+        ShrinkHelper.exportDropinAppToServer(server, war, SERVER_ONLY);
+
+        server.startServer();
+    }
+
+    @Test
+    public void testEmittingNackPayload() throws Exception {
+        server.setMarkToEndOfLog();
+        int port = server.getHttpDefaultPort();
+        URL url = new URL("http://localhost:"+port+"/"+APP_NAME+"/payload");
+        HttpURLConnection conn = HttpUtils.getHttpConnection(url, HttpUtils.DEFAULT_TIMEOUT, HttpUtils.HTTPRequestMethod.POST);
+        OutputStream os = conn.getOutputStream();
+        os.write(("test").getBytes());
+        os.flush();
+        assertThat(conn.getResponseCode(), is(500));
+        os.close();
+        conn.disconnect();
+        // check that we get error about the payload emitter channel
+        assertNotNull("Channel can ",server.waitForStringInLogUsingMark("CWMRX1003E.*restful-emitter-payload"));
```

Comment:
This is a bit of a faff isn't it.

Tempted to suggest we should try-catch around the IO to ensure streams are closed, but in a test like this it's _probably_ not a big problem.

If we end up doing this a lot, I might be tempted to extend `componenttest.topology.utils.HttpRequest` to add more general support for POST requests, but otherwise it can be left as-is.

Reply:
~Another option would be to change the rest resource to not use POST and then use `HttpRequest` as-is.~

I see in the next test why you need to pass in a body.